stellargraph/core/graph.py

```diff
@@ -331,9 +346,147 @@ class StellarGraphBase:
+        """
+        Obtains the collection of edges in the graph.
+
+        Returns:
+            The graph edges.
+        """
+        raise NotImplementedError
+
+    def has_node(self, node: Any) -> bool:
+        """
+        Indicates whether or not the graph contains the specified node.
+
+        Args:
+            node (any): The node.
+
+        Returns:
+             bool: A value of True (cf False) if the node is
+             (cf is not) in the graph.
+        """
+        raise NotImplementedError
+
+    def neighbour_nodes(self, node: Any) -> Iterable[Any]:
+        """
+        Obtains the collection of neighbouring nodes connected
+        to the given node.
+
+        Args:
+            node (any): The node in question.
+
+        Returns:
+            iterable: The neighbouring nodes.
+        """
+        raise NotImplementedError
+
+    def in_nodes(self, node: Any) -> Iterable[Any]:
+        """
+        Obtains the collection of neighbouring nodes with edges
+        directed to the given node. For an undirected graph,
+        neighbours are treated as both in-nodes and out-nodes.
+
+        Args:
+            node (any): The node in question.
+
+        Returns:
+            iterable: The neighbouring in-nodes.
+        """
+        raise NotImplementedError
+
+    def out_nodes(self, node: Any) -> Iterable[Any]:
+        """
+        Obtains the collection of neighbouring nodes with edges
+        directed from the given node. For an undirected graph,
+        neighbours are treated as both in-nodes and out-nodes.
+
+        Args:
+            node (any): The node in question.
+
+        Returns:
+            iterable: The neighbouring out-nodes.
+        """
+        raise NotImplementedError
+
+    ##################################################################
+    # Computationally intensive methods:
+
+    def node_degrees(self) -> Mapping[Any, int]:
+        """
+        Obtains a map from node to node degree.
+
+        Returns:
+            The degree of each node.
+        """
+        raise NotImplementedError
+
+    def adjacency_weights(self):
+        """
+        Obtains a SciPy sparse adjacency matrix of edge weights.
+
+        Returns:
+             The weighted adjacency matrix.
+        """
+        raise NotImplementedError
+
+
+class StellarDiGraph(StellarGraph):
+    pass
+
+
+class NetworkXStellarGraph(StellarGraph):
+    """
+    Implementation based on encapsulating a NetworkX graph.
+    """
+
+    def __init__(self, graph=None, is_directed=False, **attr):
+        if is_directed:
+            if not isinstance(graph, nx.MultiDiGraph):
+                graph = nx.MultiDiGraph(graph)
+        else:
+            if not isinstance(graph, nx.MultiGraph):
+                graph = nx.MultiGraph(graph)
+        self._graph = graph
+
+        # Name of optional attribute for edge weights
+        self._edge_weight_label = attr.get("edge_weight_label", "weight")
```

Comment:
The existing use-cases seem to pick exactly one attribute, typically `weight` by default, to be the edge weight. I agree that there can be multiple types of edge weights, but if these exist at the same time, I imagine they would become edge features (not yet implemented). Again, I chose the simplest interface to match actual usage. I have no objection to this changing in future if necessary.

Reply:
Ok, line 489 and down (`edge_weight_label`, etc.) were existing code, I think (been a while). What part of this has moved into StellarGraph? Ok, maybe you are talking about why this has to be defined up front instead of passed to each algorithm individually? Future creep, I guess.

A fundamental assertion of the subsequent branch (which includes the new `StandardStellarGraph` implementation) is that node data and edge data need to be fully encapsulated up front into standardised interfaces. Whilst it would be possible to include a method giving back data for an arbitrary 'column', I instead made it part of the design to pre-designate all necessary columns.

Based on the usages I have seen, it didn't seem problematic to just have different StellarGraph instances if one wanted to experiment with different edge weight variables. We can rethink this if it is or becomes a major decision point.

Reply:
> What part of this has moved into StellarGraph? Ok, maybe you are talking about why this has to be defined up front instead of passed to each algorithm individually? 

Yes, this `edge_weight_label` argument and the `self._edge_weight_label` processing has moved from each algorithm into the `StellarGraph` class.

> A fundamental assertion of the subsequent branch (which includes the new StandardStellarGraph implementation) is that node data and edge data need to be fully encapsulated up front into standardised interfaces

Nodes support arbitrary attributes in some sort of standardised interface, I imagine?

> Whilst it would be possible to include a method giving back data for an arbitrary 'column', I instead made it part of the design to pre-designate all necessary columns.

I don't think edge weight is a necessary column (it doesn't seem to even be necessary for most of the algorithms that support obeying weights). How does this model an unweighted graph? Does it force the user to manually specify `"weight": 1` for every edge?

From a neon/platform perspective, the more generic approach of arbitrary edge attributes is the right one: e.g. the bitcoin graph has timestamps on the edges, which don't make sense to pretend to be weights (and, some feature engineering uses these timestamps, although that processing is currently done purely in spark).

Reply:
My plan is to implement edge features along the lines of node features (see #520). Whether or not this solves the problem of 'arbitrary' columns, I don't know. By default, if you don't specify an edge weight, it is usually assumed to be 1 - typically this would be a virtual value.

If we can come up with some simple use-cases for different sorts of edge information, then I am happy to have a (group) rethink about the design. I've been purely guessing so far.

Reply:
> Whether or not this solves the problem of 'arbitrary' columns, I don't know

I would think so; in my mind, edge features and columns of data on edges are synonymous (but differently presented views) of the same concept?

> By default, if you don't specify an edge weight, it is usually assumed to be 1 - typically this would be a virtual value

I think the piece I was missing is that all of the downstream code handles this defaulting already (given I didn't notice any changes along these lines already), e.g. `to_scipy_sparse_matrix` handles seems to ignore edges that don't have the specified `weight` property, and `BiasedRandomWalk` seems to just branch on `if weighted:` at a very early point.

> If we can come up with some simple use-cases for different sorts of edge information, then I am happy to have a (group) rethink about the design. I've been purely guessing so far.

One example: the feature engineering done on bitcoin uses edge attributes; you can see some of this done in PySpark in https://github.com/stellargraph/neon/blob/develop/scenarios/scenario-2/Bitcoin_1_Features_Engineering.ipynb in "Add transaction features" where it aggregates the transactions (edges) per day, using the timestamp associated with those edges.